### PR TITLE
Allow enregistering more structs args

### LINKED ISSUE #37924: Enable promotion of structs passed in multiple registers with matching fields
With  #36862 we have support for enregistering structs that are returned in multiple registers, but structs that are passed in multiple registers are still sometimes disabled from being enregistered, even if the number of fields matches the number of registers. This includes at least some HFAs. In particular, `Compiler::StructPromotionHelper::ShouldPromoteStructVar()` will only return true for a parameter if there are 1 or 2 fields, and will return false on x64/ux for some cases of 2 fields where the offset of the second field is not 8 bytes.

category:cq
theme:structs
skill-level:expert
cost:medium

### PR DESCRIPTION
**Pull request by CarolEidt**
Allow HFAs and other structs with matching fields and registers.

Fix #37924

### CODE REVIEW

**Comment by CarolEidt** [comment]
This results in improvements primarily for Arm targets. Here are the crossgen diffs for frameworks & benchmarks (using altjits for x64UX, Arm and Arm64):

| Target | Smaller | Larger  | Size Diff | Faster  | Slower  | PerfScore Diff |
|--------|---------|---------|-----------|---------|---------|-----------|
| X64Ux  | 4       | 0       | -300 (-0.000%)   | 4   | 0   | -91.10 (-0.000%)    |
| Arm64  | 2409    | 17      | -45880 (-0.044%) | 2426 | 11 | -18290.00 (-0.002%) |
| Arm32  | 300     | 6       | -9848 (-0.014%)  |  301 | 6  | -3762.80 (-0.000%)  |

The regressions result primarily from register allocation issues (we don't prefer callee-save for floating point registers, so promoting a floating point field is not always profitable if it's live across a call).

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL
The first commit is #39284, which is still open (waiting for some test re-runs).

File: src/coreclr/src/jit/lclvars.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/lclvars.cpp:5169]:
```suggestion
//    varDsc - the local variable descriptor?
```

File: src/coreclr/src/jit/lclvars.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/lclvars.cpp:5461]: Do we leave it set to `BAD_STK_OFFS` after this change?

File: src/coreclr/src/jit/morph.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/morph.cpp:17408]: I think I am not following it. Could you please explain?  `structPromotionHelper` has `structPromotionInfo` that is a private temporary buffer that we use to keep information about the last analyzed type. We access it only in `TryPromoteStructVar` and nowhere else. 
Where could we access  `structPromotionHelper->structPromotionInfo.typeHnd` that is cleared here?

File: src/coreclr/src/jit/codegencommon.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/codegencommon.cpp:4506]: Nit: extract `regArgTab[nextArgNum]` to avoid the line break and use it in the next line as well.

**Review by sandreenko — DISMISSED** [review]
It looks like a risky change, but I understand that it is better to get it into the Prev rather that an RC.
I have a few small question and hope you could run stress tests before the merge.
LGTM in case I don't wake up before the snap.

File: src/coreclr/src/jit/lclvars.cpp
Comment (CarolEidt) [reply src/coreclr/src/jit/lclvars.cpp:2278]: I'm not following. The `compiler->lvaIsImplicitByRefLocal(lclNum)` case is handled in the `else` clause for the `#if FEATURE_MULTIREG_ARGS && defined(FEATURE_SIMD)` case. In all cases, we will call `SetArgReg()` unless I've missed something.

Comment (sandreenko) [reply src/coreclr/src/jit/lclvars.cpp:2278]: You are right, I have mixed up `#endif` with `#else` probably, thanks for the explanation.

File: src/coreclr/src/jit/lclvars.cpp
Comment (CarolEidt) [reply src/coreclr/src/jit/lclvars.cpp:1868]: It may, though I haven't run into them. That, however, is an altjit-specific issue so I don't think it should hold up this PR.

File: src/coreclr/src/jit/lclvars.cpp
Comment (CarolEidt) [reply src/coreclr/src/jit/lclvars.cpp]: This comment was reworded, hopefully it's more clear now.

File: src/coreclr/src/jit/lclvars.cpp
Comment (CarolEidt) [reply src/coreclr/src/jit/lclvars.cpp:5462]: No, this was duplicative; the field offsets are set in the promoted struct case, and don't need to be set again when they are encountered.

File: src/coreclr/src/jit/morph.cpp
Comment (CarolEidt) [reply src/coreclr/src/jit/morph.cpp:17528]: The cache is used by the inliner, during which it is conservative about looking up SIMD info. I clarified the comment a bit.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL
There are two new commits:
- https://github.com/dotnet/runtime/pull/39326/commits/5592adf4f5fbfdf600f0f4fc436c45deabcfe491 which fixes a failure on x64 Linux.
- https://github.com/dotnet/runtime/pull/39326/commits/a05526c8eae86937faf39899c0ff8a595975bbef which addresses a previous PR feedback.

Prior to the latest push, I ran jitStressRegs, and there were some test build failures. There were no actual test failures, but having rebased, once the normal CI tests complete I'll run jitStressRegs again and hopefully the test build will succeed.

**Comment by CarolEidt** [comment]
ping @dotnet/jit-contrib 
This has now passed both the regular CI and jitstressregs

**Comment by AndyAyersMS** [comment]
Does this fix the lack of enregistration/promotion in `InsertionSort` seen over on #41741 (`Span` passed by-value)?

It would show up in PMI diffs, I think.

File: src/coreclr/src/jit/morph.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/morph.cpp:17516]: Thanks for the explanation and the comment, I would probably rephrase it as: `since it is used during importation for inlining decisions.`

the place that uses it before we call struct promotion is:
https://github.com/dotnet/runtime/blob/334bd8491f17884536b8d64da4c2e36a50517ee4/src/coreclr/src/jit/importer.cpp#L18651-L18658

**Review by sandreenko — APPROVED** [review]
LGTM

**Comment by CarolEidt** [comment]
@AndyAyersMS - this doesn't fix the `InsertionSort` case. For some reason the parameter is still marked `do-not-enreg[SF] ld-addr-op`; I haven't analyzed it to determine why.

Here are the diffs:
| Arch  | OS      | What              | Delta            | Methods Improved | Methods Regressed |
| ----  | :--     | :--               | --:              | --:              | --:               |
| Arm64 | Windows | Crossgen fx+tests | -6676 (-0.01%) | 307 | 13 |
| x64   | Linux   | Crossgen fx+tests | -1781 (-0.00%) | 57 | 0 |
| Arm64 | Windows | PMI fx+benchmarks | -9952 (-0.02%) | 447 | 4 |
| x64   | Linux   | PMI fx+benchmarks | -3837 (-0.01%) | 136 | 20 |

I'll have a look at `InsertionSort` before merging.

**Comment by CarolEidt** [comment]
It turns out that these changes only support enregistering HFAs for Arm64; other multi-reg structs aren't promoted. I have a change that handles them, but will submit a separate PR for that.

**Pull request merged**
